### PR TITLE
Day15

### PR DESCRIPTION
**Pull request by JustineSara**

### CODE REVIEW

File: src/t/d15.clj [hunk 1]
```diff
@@ -0,0 +1,298 @@
...
+                  (cljstr/replace "\n" ""))]
+  [warehouse moves]))
+
+(defn symb-to-dir
+  [symb]
+  (case symb
+    \> [+1 0]
+    \< [-1 0]
+    \v [0 +1]
+    \^ [0 -1]))
```
Comment (gaverhae) [inline src/t/d15.clj]: "Symbol" is a specific type in Clojure; here you are dealing with characters.
```suggestion
(defn char-to-dir
  [c]
  (case c
    \> [+1 0]
    \< [-1 0]
    \v [0 +1]
    \^ [0 -1]))
```

File: src/t/d15.clj [hunk 2]
```diff
@@ -0,0 +1,298 @@
...
+
+(defn robot-move-box
+  [robot box1 dir box walls]
+  (loop [might-move #{}
+         try-move box1
+         boxes box]
+    (let [npos (map + try-move dir)]
+      (cond
+        (contains? boxes npos)
+        (recur (cljset/union might-move #{try-move}) npos (disj boxes npos))
```
Comment (gaverhae) [inline src/t/d15.clj:94]: Why not `conj`? Am I missing how this is different?
```suggestion
        (recur (conj might-move try-move) npos (disj boxes npos))
```

File: src/t/d15.clj [hunk 3]
```diff
@@ -0,0 +1,298 @@
...
+        (contains? walls npos)
+        [robot (cljset/union might-move #{try-move} boxes)]
+        :else
```
Comment (gaverhae) [inline src/t/d15.clj]: Note that `else` is not special here, any non-falsy value will do. Sometimes you can make your code more readable by either explicitly stating the (arguably redundant) condition, as in
```clojure
(cond (> a b) ...
      (= a b) ...
      (< a b) ...)
```
and sometimes you can use a different keyword, or even a string, to explain the last case more explicitly:
```suggestion
        "We have found an empty space, all boxes tried so far can move."
```

File: src/t/d15.clj [hunk 4]
```diff
@@ -0,0 +1,298 @@
...
+             (cljset/union (cljset/difference box-r mov-box-r) (set (map (fn[[x y]] [(inc x) y]) mov-box-r)))])))
+
+(defn move-left
+  [[rx ry] box-l box-r walls]
+  (loop [testx (dec rx)
+         mov-box-l #{}
+         mov-box-r #{}
+         ]
+    (cond
+      (contains? box-r [testx ry]) (recur (- testx 2) (cljset/union mov-box-l #{[(dec testx) ry]}) (cljset/union mov-box-r #{[testx ry]}))
```
Comment (gaverhae) [inline src/t/d15.clj:187]: If you're intent on making `cljset/union` your one true operation, note that you can do something like

```clojure
(def ∪ cljset/union)
```

to make the rest of the file a little bit more terse (if you think that's a good idea).

File: src/t/d15.clj [hunk 5]
```diff
@@ -0,0 +1,298 @@
...
+  (let [dir (symb-to-dir symb)
+        n-robot (map + robot dir)]
+    ;; try to move the robot
+    (cond
+      (contains? walls n-robot) [robot box-l box-r]
+      (contains? (cljset/union box-l box-r) n-robot) (case symb
+                                                       \> (move-right robot box-l box-r walls)
+                                                       \< (move-left  robot box-l box-r walls)
+                                                       \v (move-down  robot box-l box-r walls)
+                                                       \^ (move-up    robot box-l box-r walls))
```
Comment (gaverhae) [inline src/t/d15.clj]: `case` is an expression and functions are first-class, so you could write this as:

```suggestion
      (contains? (cljset/union box-l box-r) n-robot) ((case symb
                                                        \> move-right
                                                        \< move-left
                                                        \v move-down
                                                        \^ move-up) robot box-l box-r walls))
```

**Review by gaverhae — APPROVED** [review]

File: src/t/d15.clj (hunk 2 above)
Comment (JustineSara) [reply src/t/d15.clj:94]: I lost the set form at some point, I thought `conj` was/could-be responsible? (I just retested it and apparently I keep the set thing so ... I don't know)

File: src/t/d15.clj (hunk 4 above)
Comment (JustineSara) [reply src/t/d15.clj:187]: I really had an issue with sets becoming not-sets and so I used only things I was certain would return a set
I did not start the day thinking "I will need union" (pun intended TT.TT)

File: src/t/d15.clj (hunk 2 above)
Comment (gaverhae) [reply src/t/d15.clj:94]: Functions in clojure.core are, broadly speaking, either "collection" functions or seq ("sequence") functions. Collection functions preserve their input type, whereas seq functions always convert their input to a seq first (typically by calling `seq`) and thus "lose" the type information.

As a rough list off the top of my head:
- `conj`, `disj`, `assoc`, `dissoc`, `count`, `peek`, `pop` are "collection" funcions.
- `first`, `rest`, `last`, `butlast`, `next`, `map`, `filter`, `reduce`, `remove`, `group-by`, `sort` are sequence functions.

It's a very important distinction that may not be all that apparent if you've not been told about it. I don't think there's a way around memorizing which category a function belongs to as part of learning that function.

Comment (JustineSara) [reply src/t/d15.clj:94]: Thanks, I guess I will keep testing until I know it. But it feels like there is a sense underneath all

**Pull request merged**